### ISSUE TITLE
This repo is mentioned for classification only, so it cant be used to semantic segmentation yet?

### ISSUE THREAD

**Issue by shrutishrestha** [csrhddlam/axial-deeplab#10]
At the last layer of Axial attention, there is the implementation of the Linear fc layer, which gives the classification output. But, I need to work on semantic segmentation. So usingn axial attention as encoder, should I use this for unet decoder or else how can I proceed for segmentation task?

**Comment by csrhddlam** [comment]
Right. It does not support segmentation yet.

I would suggest to use it as an encoder. Please have a look at section 3.2 axial-deeplab for details. If you would like to use it as a decoder also, Appendix B provides our decoder design and the performance of it.

**Comment by shrutishrestha** [comment]
keeping this as encoder can other decoders (eg unet ) be used for performing semantic segmentation?

**Comment by csrhddlam** [comment]
Yes.

**Comment by shrutishrestha** [comment]
Are you planning to upload the code for the axial decoder anytime soon?

**Comment by csrhddlam** [comment]
No, but you could convert a block to a decoder block easily by inserting one or two upsamplings and 1x1 convolutions.

**Comment by shrutishrestha** [comment]
<img width="1198" alt="Screen Shot 2020-09-16 at 12 44 58 PM" src="https://user-images.githubusercontent.com/33342277/93302962-7a205980-f81a-11ea-9f3f-821c0e30613d.png">
I am getting this error. MY N=8, group planes =8/2=4, H-64, W-64

what might be the possible reason for this?

**Comment by phj128** [comment]
It seems that the sizes of q and q_embedding do not match. Could you please provide more information? Did you make any changes of the code?

**Comment by shrutishrestha** [comment]
Can you please check if this network design is correct or not while using axial deeplab encoder with decoder. Also after the decoder reaches 1,32,56,56 which are N,C,H,W what can be done to produce the output as exact size as of its image ?



![IMG_20200917_085523__01.jpg](https://user-images.githubusercontent.com/33342277/93415644-b52a9780-f8c3-11ea-8c75-cc3879554f73.jpg)

**Comment by phj128** [comment]
It seems OK

> Can you please check if this network design is correct or not while using axial deeplab encoder with decoder. Also after the decoder reaches 1,32,56,56 which are N,C,H,W what can be done to produce the output as exact size as of its image ?
> 
> ![IMG_20200917_085523__01.jpg](https://user-images.githubusercontent.com/33342277/93415644-b52a9780-f8c3-11ea-8c75-cc3879554f73.jpg)

**Comment by shrutishrestha** [comment]
In the decoder block, we go 3,4,6,3 blocks in each layer as in encoder side, or else we have only one block in each layer as shown in figure above?

**Comment by sahilrider** [comment]
@shrutishrestha Have you completed the decoder block?

**Comment by sahilrider** [comment]
In the paper it is written "_Firstly, to extract dense feature maps, DeepLab [13] changes the stride and
atrous rates of the last one or two stages in ResNet [32]. Similarly, we remove the
stride of the last stage but we do not implement the `atrous' attention module,
since our axial-attention already captures global information for the whole input._"

Does this mean for the last layer we don't need to do the downsampling and the output size of the last layer will be [1, 1024, 14, 14]? @csrhddlam

**Comment by csrhddlam** [comment]
Right, it will be [1, 1024, 14, 14], although in practice, the input resolution for semantic segmentation is usually larger than 224x224.

**Comment by sahilrider** [comment]
Yes right. I was assuming 224x224 for calculation. Thanks.

**Comment by lkqnaruto** [comment]
> Can you please check if this network design is correct or not while using axial deeplab encoder with decoder. Also after the decoder reaches 1,32,56,56 which are N,C,H,W what can be done to produce the output as exact size as of its image ?
> 
> ![IMG_20200917_085523__01.jpg](https://user-images.githubusercontent.com/33342277/93415644-b52a9780-f8c3-11ea-8c75-cc3879554f73.jpg)

I see the skip connection will be passed through conv1x1, would you please explain why you design in this way?

The second thing I notice is that in the decoder (in the right down corner), the input size is 1x1024x7x7, but somehow it becomes 1x512x7x7 before conv and upsample without any operation?

Thank you
